wallet/psbt.go

```diff
@@ -0,0 +1,300 @@
+		// When checking an output for things like dusty-ness, we'll
+		// use the default mempool relay fee rather than the target
+		// effective fee rate to ensure accuracy. Otherwise, we may
+		// mistakenly mark small-ish, but not quite dust output as
+		// dust.
+		err := txrules.CheckOutput(output, txrules.DefaultRelayFeePerKb)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Let's find out the amount to fund first.
+	amt := int64(0)
+	for _, output := range txOut {
+		amt += output.Value
+	}
+
+	// addInputInfo is a helper function that fetches the UTXO information
+	// of an input and attaches it to the PSBT packet.
+	addInputInfo := func(inputs []*wire.TxIn) error {
+		packet.Inputs = make([]psbt.PInput, len(inputs))
+		for idx, in := range inputs {
+			utxo, _, err := w.FetchInputInfo(&in.PreviousOutPoint)
+			if err != nil {
+				return fmt.Errorf("error fetching UTXO: %v",
+					err)
+			}
+
+			packet.Inputs[idx].WitnessUtxo = &wire.TxOut{
+				Value:    utxo.Value,
+				PkScript: utxo.PkScript,
+			}
+			packet.Inputs[idx].SighashType = txscript.SigHashAll
+
+			// We don't want to include the witness just yet.
+			packet.UnsignedTx.TxIn[idx].Witness = wire.TxWitness{}
+		}
+
+		return nil
+	}
+
+	var tx *txauthor.AuthoredTx
+	switch {
+	// We need to do coin selection.
+	case len(txIn) == 0:
+		// We ask the underlying wallet to fund a TX for us. This
+		// includes everything we need, specifically fee estimation and
+		// change address creation.
+		tx, err = w.CreateSimpleTx(
+			account, packet.UnsignedTx.TxOut, 1, feeSatPerKB,
+			false,
+		)
+		if err != nil {
+			return fmt.Errorf("error creating funding TX: %v", err)
+		}
+
+		// Copy over the inputs now then collect all UTXO information
+		// that we can and attach them to the PSBT as well. We don't
+		// include the witness as the resulting PSBT isn't expected not
+		// should be signed yet.
+		packet.UnsignedTx.TxIn = tx.Tx.TxIn
+		err = addInputInfo(tx.Tx.TxIn)
+		if err != nil {
+			return err
+		}
+
+	// If there are inputs, we need to check if they're sufficient and add
+	// a change output if necessary.
+	default:
+		// Make sure all inputs provided are actually ours.
+		err = addInputInfo(txIn)
+		if err != nil {
+			return err
+		}
+
+		// We can leverage the fee calculation of the txauthor package
+		// if we provide the selected UTXOs as a coin source.
+		credits := make([]wtxmgr.Credit, len(txIn))
+		for idx, in := range txIn {
+			utxo := packet.Inputs[idx].WitnessUtxo
+			credits[idx] = wtxmgr.Credit{
+				OutPoint: in.PreviousOutPoint,
+				Amount:   btcutil.Amount(utxo.Value),
+				PkScript: utxo.PkScript,
+			}
+		}
+		inputSource := makeInputSource(credits)
+
+		// We also need a change source which needs to be able to insert
+		// a new change addresse into the database.
+		dbtx, err := w.db.BeginReadWriteTx()
+		if err != nil {
+			return err
+		}
+		addrmgrNs := dbtx.ReadWriteBucket(waddrmgrNamespaceKey)
+		changeSource := func() ([]byte, error) {
```

Comment:
I believe this logic also exists elsewhere, can we refactor it into a method and use it here?